sklearn/preprocessing/target.py

```diff
@@ -0,0 +1,222 @@
+    >>> tt.fit(X, y)
+    ... #doctest: +NORMALIZE_WHITESPACE
+    TransformedTargetRegressor(check_invertible=True,
+        estimator=LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1,
+                                   normalize=False),
+        func=<ufunc 'log'>, inverse_func=<ufunc 'exp'>, transformer=None)
+    >>> tt.score(X, y)
+    1.0
+    >>> tt.estimator_.coef_
+    array([ 2.])
+
+    """
+    def __init__(self, estimator=LinearRegression(), transformer=None,
+                 func=None, inverse_func=None, check_invertible=True):
+        self.estimator = estimator
+        self.transformer = transformer
+        self.func = func
+        self.inverse_func = inverse_func
+        self.check_invertible = check_invertible
+        # we probably need to change this ones we have tags
+        self._estimator_type = estimator._estimator_type
+
+    def _validate_transformer(self, y):
+        if (self.transformer is not None and
+                (self.func is not None or self.inverse_func is not None)):
+            raise ValueError("Both 'transformer' and functions 'func'/"
+                             "'inverse_func' cannot be set at the same time.")
+        elif self.transformer is not None:
+            self.transformer_ = clone(self.transformer)
+        else:
+            self.transformer_ = FunctionTransformer(
+                func=self.func, inverse_func=self.inverse_func, validate=False)
+        self.transformer_.fit(y)
+        if self.check_invertible:
+            n_subsample = min(1000, y.shape[0])
+            subsample_idx = np.random.choice(range(y.shape[0]),
+                                             size=n_subsample, replace=False)
+            diff = np.abs((y[subsample_idx] -
+                           self.transformer_.inverse_transform(
+                               self.transformer_.transform(y[subsample_idx]))))
+            if np.sum(diff) > 1e-7:
+                raise ValueError("The provided functions or transformer are"
+                                 " not strictly inverse of each other.")
+
+    def fit(self, X, y, sample_weight=None):
+        """Fit the model according to the given training data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Training vector, where n_samples is the number of samples and
+            n_features is the number of features.
+
+        y : array-like, shape (n_samples,)
+            Target vector relative to X.
+
+        sample_weight : array-like, shape (n_samples,) optional
+            Array of weights that are assigned to individual samples.
+            If not provided, then each sample is given unit weight.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        # memorize if y should be a multi-output
+        self.y_ndim_ = y.ndim
+        if y.ndim == 1:
+            y_2d = y.reshape(-1, 1)
```

Comment:
I come back on this point. I am not sure this is a great idea since that it changes the behaviour between if you pass a function or a transformer. We could still make the transform to 2d and build the FunctionTransformer with `validate=True` and the behaviour will always be the same.

I don't see a case in which the user would define a function which working on a 1D array but would failed on a 2D array